[SPARK-34482][SS] Correct the active SparkSession for StreamExecution.logicalPlan

Ngone51 opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

Set the active SparkSession to `sparkSessionForStream` and diable AQE & CBO before initializing the `StreamExecution.logicalPlan`.

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->

The active session should be `sparkSessionForStream`. Otherwise, settings like

https://github.com/apache/spark/blob/6b34745cb9b294c91cd126c2ea44c039ee83cb84/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala#L332-L335

wouldn't take effect if callers access them from the active SQLConf, e.g., the rule of `InsertAdaptiveSparkPlan`. Besides, unlike `InsertAdaptiveSparkPlan` (which skips streaming plan), `CostBasedJoinReorder` seems to have the chance to take effect theoretically.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->

No.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

Tested manually. Before the fix, `InsertAdaptiveSparkPlan` would try to apply AQE on the plan(wouldn't take effect though). After this fix, the rule returns directly.

Ngone51 committed:
fix

Ngone51 commented:
cc @viirya @xuanyuanking @HeartSaVioR Please take a look, thanks!

SparkQA commented:
**[Test build #135306 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135306/testReport)** for PR 31600 at commit [`52a6999`](https://github.com/apache/spark/commit/52a6999d3c602f4968a3f1e2d137e1cb5d25cd4e).

viirya reviewed (approved):
looks correct.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39886/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/39886/

HeartSaVioR commented:
Just wondering, does anyone know the reason the code uses both `sparkSession` and `sparkSessionForStream`? Are there any known conditions when we should still use `sparkSession` and when we should use `sparkSessionForStream` instead?

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/39886/

SparkQA commented:
**[Test build #135306 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135306/testReport)** for PR 31600 at commit [`52a6999`](https://github.com/apache/spark/commit/52a6999d3c602f4968a3f1e2d137e1cb5d25cd4e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135306/

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala:
```diff
@@ -323,7 +323,7 @@ abstract class StreamExecution(
 
       // While active, repeatedly attempt to run batches.
-      SparkSession.setActiveSession(sparkSession)
+      SparkSession.setActiveSession(sparkSessionForStream)
```
Can we have a test case by using your PR description, @Ngone51 ?

Ngone51 replied:
ok, let me try it.

HyukjinKwon reviewed (approved):
Looks good. Having a test should be good if it's not too difficult or large

Ngone51 commented:
> Just wondering, does anyone know the reason the code uses both sparkSession and sparkSessionForStream? Are there any known conditions when we should still use sparkSession and when we should use sparkSessionForStream instead?


That's a good question. After taking another look, IIUC, I think `sparkSessionForStream`  is used for  `runActivatedStream()` only and `sparkSession` is used for everything  out of `runActivatedStream()`.  Thus, I think the better fix should be:

```scala
sparkSessionForStream.withActive {
 ...
 runActivatedStream(sparkSessionForStream)
 ...
}
```

What do you guys think?

HeartSaVioR commented:
Given we assume that the code might have a bug, it may not be safe if we try to answer the question based on current code. Probably we'd need to try to ask and figure about the "historical" information what sparkSessionForStream is for, and why sparkSession should be co-used even after cloning the session.

cc. @mukulmurthy @zsxwing @jose-torres who authored/reviewed SPARK-26586. also @tdas who is likely having context.

zsxwing commented:
I don't remember why we didn't set the active session to `sparkSessionForStream`. Probably just overlook when we introduced the cloned session for streaming queries.

@Ngone51 could you add a unit test for this?

Ngone51 commented:
> @Ngone51 could you add a unit test for this?

Yeah, I'm trying...

Ngone51 committed:
disable aqe/cbo before init logicalPlan

Ngone51 committed:
add test

Ngone51 commented:
Hi all, I realized that the issue actually only affect the `StreamExecution.logicalPlan`. (I have added the test for it).

While for the `runActivatedStream(sparkSessionForStream)`, the active session has been correctly set to `sparkSessionForStream`.

I have updated the PR, so please take another look.

Ngone51 commented on sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala:
```diff
@@ -323,26 +323,28 @@ abstract class StreamExecution(
-        updateStatusMessage("Stopped")
-      } else {
-        // `stop()` is already called. Let `finally` finish the cleanup.
+      sparkSessionForStream.withActive {
```
Using `withActive` to keep safe for the `StreamExecution` to avoid using the wrong active session.

SparkQA commented:
**[Test build #135472 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135472/testReport)** for PR 31600 at commit [`ddbeca7`](https://github.com/apache/spark/commit/ddbeca731c1219dbeb96de64840e80b7a4b715c7).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40052/

SparkQA commented:
**[Test build #135472 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135472/testReport)** for PR 31600 at commit [`ddbeca7`](https://github.com/apache/spark/commit/ddbeca731c1219dbeb96de64840e80b7a4b715c7).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135472/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40052/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/40052/

cloud-fan commented:
Looks reasonable to me. I'll merge it within a few days if there is no objection.

Ngone51 commented:
retest this please

SparkQA commented:
**[Test build #135587 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135587/testReport)** for PR 31600 at commit [`ddbeca7`](https://github.com/apache/spark/commit/ddbeca731c1219dbeb96de64840e80b7a4b715c7).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40168/

SparkQA commented:
Kubernetes integration test status failure
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/40168/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/40168/

SparkQA commented:
**[Test build #135587 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/135587/testReport)** for PR 31600 at commit [`ddbeca7`](https://github.com/apache/spark/commit/ddbeca731c1219dbeb96de64840e80b7a4b715c7).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/135587/

HeartSaVioR reviewed (approved):
+1

As I've said, it'd be nice if we could clarify `sparkSession` vs `sparkSessionForStream`, but seems like we don't have the context of history, so doesn't seem to be easy to do that. New UT looks to clarify the needs of the PR.

xuanyuanking reviewed (approved):
LGTM

cloud-fan commented:
thanks, merging to master/3.1!

dongjoon-hyun commented:
Thank you so much, @Ngone51 and all!

Ngone51 commented:
thanks all!
